overlord/snapstate/autorefresh_gating.go

```diff
@@ -33,6 +36,270 @@ import (
+		return d1
+	}
+	return d2
+}
+
+// HoldRefresh marks affectingSnaps as held for refresh for up to holdTime.
+// HoldTime of zero denotes maximum allowed hold time.
+// Holding may fail for only some snaps in which case HoldError is returned and
+// it contains the details of failed ones.
+func HoldRefresh(st *state.State, gatingSnap string, holdDuration time.Duration, affectingSnaps ...string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	herr := &HoldError{
+		SnapsInError: make(map[string]HoldDurationError),
+	}
+	now := timeNow()
+	for _, heldSnap := range affectingSnaps {
+		hold, ok := gating[heldSnap][gatingSnap]
+		if !ok {
+			hold = &holdState{
+				FirstHeld: now,
+			}
+		}
+
+		lastRefreshTime, err := lastRefreshed(st, heldSnap)
+		if err != nil {
+			return err
+		}
+
+		mp := maxPostponement - maxPostponementBuffer
+		maxDur := maxAllowedPostponement(gatingSnap, heldSnap, mp)
+
+		// calculate max hold duration that's left considering previous hold
+		// requests of this snap and last refresh time.
+		left := holdDurationLeft(now, lastRefreshTime, hold.FirstHeld, maxDur, mp)
+		if left <= 0 {
+			herr.SnapsInError[heldSnap] = HoldDurationError{
+				Err: fmt.Errorf("snap %q cannot hold snap %q anymore, maximum refresh postponement exceeded", gatingSnap, heldSnap),
+			}
+			continue
+		}
+
+		dur := holdDuration
+		if dur == 0 {
+			// duration not specified, using a default one (maximum) or what's
+			// left of it.
+			dur = left
+		} else {
+			// explicit hold duration requested
+			if dur > maxDur {
+				herr.SnapsInError[heldSnap] = HoldDurationError{
+					Err:          fmt.Errorf("requested holding duration %s exceeds maximum holding for snap %q", holdDuration, heldSnap),
+					DurationLeft: left,
+				}
+				continue
+			}
+		}
+
+		newHold := now.Add(dur)
+		cutOff := lastRefreshTime.Add(maxPostponement - maxPostponementBuffer)
+
+		// consider last refresh time and adjust hold duration if needed so it's
+		// not exceeded.
+		if newHold.Before(cutOff) {
+			hold.HoldUntil = newHold
+		} else {
+			hold.HoldUntil = cutOff
+		}
+
+		// finally store/update gating hold data
+		if _, ok := gating[heldSnap]; !ok {
+			gating[heldSnap] = make(map[string]*holdState)
+		}
+		gating[heldSnap][gatingSnap] = hold
+	}
+
+	if len(herr.SnapsInError) != len(affectingSnaps) {
+		st.Set("snaps-hold", gating)
+	}
+	if len(herr.SnapsInError) > 0 {
+		return herr
+	}
+	return nil
+}
+
+// ProceedWithRefresh unblocks all snaps held by gatingSnap for refresh. This
+// should be called for --proceed on the gatingSnap.
+func ProceedWithRefresh(st *state.State, gatingSnap string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	if len(gating) == 0 {
+		return nil
+	}
+
+	var changed bool
+	for heldSnap, gatingSnaps := range gating {
+		if _, ok := gatingSnaps[gatingSnap]; ok {
+			delete(gatingSnaps, gatingSnap)
+			changed = true
+		}
+		if len(gating[heldSnap]) == 0 {
```

Comment:
this could also just check `gatingSnaps` no?

```suggestion
		if len(gatingSnaps) == 0 {
```

Reply:
Yes, you're right, thanks!